cmd/entrypoint/syntheticauth.go

```diff
@@ -40,13 +43,70 @@ func ReconcileAuthServices(ctx context.Context, sh *SnapshotHolder, deltas *[]*k
 	syntheticAuthExists := false
 	for _, authService := range sh.k8sSnapshot.AuthServices {
 		// Keep any AuthServices already using protocol_version: v3
-		if authService.ObjectMeta.Name == "synthetic-edge-stack-auth" {
-			syntheticAuthExists = true
-		} else if authService.Spec.ProtocolVersion == "v3" {
-			authServices = append(authServices, authService)
+		if authService.Spec.ProtocolVersion == "v3" {
+			injectSyntheticAuth = false
+			if authService.ObjectMeta.Name == "synthetic-edge-stack-auth" {
+				syntheticAuthExists = true
+			} else {
+				authServices = append(authServices, authService)
+			}
 		}
 	}
-	if len(authServices) == 0 && !syntheticAuthExists {
+
+	// Also loop over the annotations and remove authservices that are not v3. We do
+	// this by looping over each entry in the annotations map, removing all the non-v3
+	// AuthService entries, and then removing any keys that end up with empty lists.
+	//
+	// keysToDelete tracks keys to delete, since I never trust deleting things from
+	// collections while iterating over them.
+	keysToDelete := make([]string, 0)
```

Comment:
@kflynn wrote this section that loops through the annotations, and then I edited the commit while cleaning up the git history. Originally he wrote it when we were thinking that these Annotations had to be treated as `kates.Unstructured`. I noticed that they were actually typed so there is some unneeded confusion since I left the comments as they were. I'll clean this section up, but thats just some context for some of the other replies I'll be leaving to provide some info/context before I push some requested changes.

As for deleting items from a collection while iterating over them, I assume Flynn just wrote it that way to be safe with deleting items while iterating over them since we do put a lock on the snapshot in `entrypoint/watcher.go K8sUpdate()` where `ReconcileAuthServices()` is called. There will not be any changes to the snapshot during the course of the function regardless of what happens to the cluster. 

Even if there were rapid config changes in the cluster, we completely finish one update check and complete the reconfigure before begnning another and they all get queued up so nothing is going to be changing in the snapshot. That said, I can still add a test for this to ensure that things are always this way, but I'm just going modify this to delete the keys during the iteration since its safe to do so in golang.